R/rectangle.R

```diff
@@ -167,104 +222,461 @@ hoist <- function(.data, .col, ..., .remove = TRUE, .simplify = TRUE, .ptype = l
 check_pluckers <- function(...) {
   pluckers <- list2(...)
 
-  is_string <- vapply(pluckers, function(x) is.character(x) && length(x) == 1, logical(1))
+  is_string <- map_lgl(pluckers, ~is.character(.x) && length(.x) == 1)
   auto_name <- names2(pluckers) == "" & is_string
 
   if (any(auto_name)) {
     names(pluckers)[auto_name] <- unlist(pluckers[auto_name])
   }
 
-  if (!is_named(pluckers)) {
-    stop("All elements of `...` must be named", call. = FALSE)
+  if (length(pluckers) > 0 && !is_named(pluckers)) {
+    abort("All elements of `...` must be named.")
   }
 
   if (vec_duplicate_any(names(pluckers))) {
-    abort("The names of `...` must be unique")
+    abort("The names of `...` must be unique.")
   }
 
+  # Standardize all pluckers to lists for splicing into `pluck()`
+  # and for easier handling in `strike()`
+  is_not_list <- !map_lgl(pluckers, vec_is_list)
+  pluckers[is_not_list] <- map(pluckers[is_not_list], vec_chop)
+
   pluckers
 }
 
+strike <- function(x, indices) {
+  if (!vec_is_list(indices)) {
+    abort("Internal error: `indices` must be a list.")
+  }
+
+  n_indices <- vec_size(indices)
+
+  if (n_indices == 0L) {
+    # Edge case corresponding to an empty plucker
+    return(x)
+  }
+
+  index <- indices[[1L]]
+  indices <- indices[-1L]
+
+  is_valid_index <-
+    (is.numeric(index) && (index <= vec_size(x))) ||
+    (is.character(index) && has_name(x, index))
+
+  if (!is_valid_index) {
+    # Nothing to do if the `pluck()` missed entirely
+    return(x)
+  }
+
+  if (n_indices == 1L) {
+    # At base index, remove it
+    if (is.numeric(index)) {
+      x <- x[-index]
+    } else if (is.character(index)) {
+      x <- x[setdiff(names(x), index)]
+    } else {
+      abort("Internal error: Only character and numeric values are valid indices.")
+    }
+  } else {
+    # Not at base index yet, continue recursion
+    x[[index]] <- strike(x[[index]], indices)
+  }
+
+  x
+}
+
 #' @export
 #' @rdname hoist
-#' @param values_to Name of column to store vector values. Defaults to `col`.
-#' @param indices_to A string giving the name of column which will contain the
-#'   inner names or position (if not named) of the values. Defaults to `col`
-#'   with `_id` suffix
-#' @param indices_include Add an index column? Defaults to `TRUE` when `col`
-#'   has inner names.
+#' @param values_to A single string representing the column name to store the
+#'   unnested values in. If multiple columns are specified in `col`, this can
+#'   also be a glue string containing `"{col}"` to provide a template for the
+#'   column names. If `NULL`, defaults to `"{col}"`.
+#' @param indices_to A string giving the name of the column which will contain
```

Comment:
```suggestion
#' @param indices_to A string giving the column name to store
```

To match `values_to`